FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+
+- A. each time when a new Dataset/DataArray is created
+- B. only when we need it (i.e., when calling `.sel()` or `indexes`)
+
+#### 2.2.4 Implicit coordinates
+
+If we want to strictly follow the data model defined in [Section 1](#1_Data_Model), we need to ensure that at least one coordinate exists for each index (generally we need to ensure that the data indexed is properly exposed as one or more coordinates), especially when index objects are given directly to the Dataset/DataArray constructor ([Section 2.2.2](#222_DatasetDataArrays_indexes_constructor_argument)). This means that every `XarrayIndex` subclass must implement a method to get or generate one or more Xarray coordinates.
+
+Additionally, we might want to generate a coordinate that is specific to the index, generally with a composite data type, like a coordinate with n-elements tuple labels for a n-levels `pandas.MultiIndex`. This should be optional, though. An `XarrayIndex` may or may not support this. This coordinate probably shouldn't be created implicitly or by default when setting the new index. Also, this might raise some issues, e.g., what would be the name of this coordinate? For a `pandas.MultiIndex` it would make sense in many (but not all) cases to choose the dimension name as the coordinate name.
+
+### 2.3 Index access
+
+#### 2.3.1 Dataset/DataArray's `indexes` property
+
+The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).
+
+#### 2.3.2 Additional Dataset/DataArray's convenient methods
+
+Using tuples as keys in the `indexes` property has the advantage of providing clear information about the order that was used to build multi-coordinate indexes. For convenience, we could have an additional property `coord_indexes` or an additional method `get_index(coord_name: str)` to easily retrieve the index object that is tied to a specific coordinate.
+
+### 2.4 Propagate indexes through operations
+
+TODO
+
+- Any operation affecting one coordinate of a multi-coordinate index -> simply drop the index or maybe allow `XarrayIndex` subclasses to optionally generate and return a new or updated index
+
+### 2.5 Using indexes for data selection
+
+TODO: `.sel()` would now accept coordinate names as indexer keys.
+
+Indexes that are wrapped as `XarrayIndex` subclasses may provide more selection capabilities than what is currently possible using Dataset/DataArray `sel()`, e.g., radius or region selection, k-nearest neighbors, etc. There's no plan here to expand Xarray's selection API. This could be extended in 3rd-party libraries, e.g., using Dataset/DataArray accessors that would reuse the indexes attached to a Dataset/DataArray object.
+
+### 2.6 Index serialization
+
+TODO
```

Comment:
I don't think we need serialization for the first proof of concept. Given that we'll force all coordinates corresponding to an index to be explicit variables, direct serialization support may not be necessary. You can simply recreate the index later if necessary.

Reply:
Agreed. It's certainly too early to address this.

I was thinking about some cases where the index may be very costly to build. We could take advantage of some index implementations to "instantly" re-create the index from its serialized internal data structures (e.g., spatialpandas' [HilbertRTree](https://github.com/holoviz/spatialpandas/blob/4d679f97d750c6a69b49143c1a21e17bf6731d5d/spatialpandas/spatialindex/rtree.py#L68), scikit-learn's `KDTree` and `BallTree` that use an array-based representation of an immutable tree structure). This improvement in performance is welcome but certainly not critical at this stage.